### ISSUE TITLE
Azure App Service Environment Variables Not Working with Webjobs

### ISSUE THREAD

**Issue by rmmcgr** [aspnet/MicrosoftConfigurationBuilders#133]
I am trying to use configuration builders on a .NET full framework (4.7.2) webjobs project. Locally everything working is fine, but after deployment to Azure App Service the app service's environment variables are not being injected anymore into the configuration.

The webjobs will not start up, because the necessary configuration strings are not present.

My app.config looks like:

```
<?xml version="1.0" encoding="utf-8"?>
<configuration>

  <configSections>
    <section name="entityFramework" type="System.Data.Entity.Internal.ConfigFile.EntityFrameworkSection, EntityFramework, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" requirePermission="false" />
	<!-- For more information on Entity Framework configuration, visit http://go.microsoft.com/fwlink/?LinkID=237468 -->
	<section name="configBuilders" type="System.Configuration.ConfigurationBuildersSection, System.Configuration, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" restartOnExternalChanges="false" requirePermission="false" />
  </configSections>
  
  <configBuilders>
    <builders>
    <add name="UserSecrets" mode="Strict" userSecretsId="my_guid_id" type="Microsoft.Configuration.ConfigurationBuilders.UserSecretsConfigBuilder, Microsoft.Configuration.ConfigurationBuilders.UserSecrets, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35" />
    <add name="Environment" type="Microsoft.Configuration.ConfigurationBuilders.EnvironmentConfigBuilder, Microsoft.Configuration.ConfigurationBuilders.Environment, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35" /></builders>
  </configBuilders>
  
  <connectionStrings configBuilders="Environment,UserSecrets">
    <add name="AzureWebJobsDashboard" connectionString="" />
    <add name="AzureWebJobsStorage" connectionString="" />
    <add name="DBContext" connectionString="" providerName="System.Data.SqlClient" />
  </connectionStrings>
  
  ... other parts of the config file ...
  
</configuration>
```

When running locally, the user secrets file is successfully read and everything works as expected. While there is obviously not such a file in the App Service, I was assuming that the use of the Environment configuration builder would see the App Service configuration an inject the necessary values.

Are these assumptions wrong and/or have I not configured/used this correctly?

**Comment by shiitake** [comment]
I'm not sure if you were able to resolve this but I'm having a similar issue in my asp.net 4.8 application using the AzureKeyVaultConfigBuilder where it doesn't recognize my environment variables. 

How are you setting the environment variables on your app service?

**Comment by smurtagh** [comment]
Also having this issue when using `AzureKeyVaultConfigBuilder` with my AppSettings. I had to move all my settings to either the key vault or my app.config because the values on the App Service are not getting injected.

**Comment by StephenMolloy** [comment]
I'm a little late to the party here. So I'm not sure how helpful this will be. I see a bit of summary information, but not enough details to really know what is going on. I think I can piece some things together and take a guess though.

@rmmcgr - You are most likely correct about UserSecrets not existing in the Azure App Service environment. If you're not, then you've checked your secrets file in with your project and published it, which would be BAD. ;)  You can continue to apply config builders to you `connectionStrings` section if you like, but it appears like that interferes with Azure App Service's ability to inject connection strings you have defined in it's 'Configuration' panel. I don't know why since I don't know how they do it. And config builders don't seem to interfere the same way with appSettings. ¯\_ (ツ)_/¯

You could go a couple different ways here. Have a config transformation remove your builders from app.config before publishing to Azure. Or add an additional 'Environment' builder that will look for the connection strings with the appropriate prefix since they will still exist in the environment even if they don't make it into the config section. [See here for those prefixes.](https://docs.microsoft.com/en-us/azure/app-service/configure-common?tabs=portal#configure-connection-strings)

@shiitake - It sounds like your issue is a little different. By chance are you deploying in a container? It's a known issue that Windows containers only set environment variables for the EntryPoint process. Which means if you're running a web load or service of some type, the windows service/IIS won't see those environment variables. If you use an IIS or ASP.Net specific container, they set the EntryPoint process to be something called 'ServiceMonitor' which should handle pushing those new environment variables to the service it is monitoring.

@smurtagh - I am not seeing the same. If this is still an issue for you, please gather simple repro steps and create a new issue so I can take a better look. Thanks.

**Comment by StephenMolloy** [comment]
So I think what is happening here is that Azure App Services (functions, web jobs, etc) does their own version of settings injection that gets thrown off a little bit when you have config builders applied. I don't work on any Azure team, so I'm just speculating on how it all works... but it looks to me like Azure tries to manually insert settings by manipulating the xml of the app.config file before starting the app. But the presence of the unexpected `configBuilders="whatever"` attribute is throwing their injection routine for a loop and it just throws its hands up and does nothing. _Even if your config builder subsequently does nothing at runtime._ Just the attribute alone seems to be enough.

This is probably something that should get filed with the Azure App Services team. There's really no reason why the presence of an attribute should throw off their pre-runtime xml injection. Or maybe there is good reason, but I'm not part of the team that would know. Anyway...

The good news is that all of those Azure settings are still there in the environment along with special prefixes to make them easy to find. So if you want to pull settings from _both_ the Azure injection mechanism just before launching the app as well as your own config builder mechanisms at runtime, you can mimic that by having config builders pull in the settings that Azure would have injected. The connection strings prefix changes based on the type of connection string, so you need a few different config builders for `<connectionStrings>` to faithfully reproduce everything that could have been injected. But it's not too difficult. Something like this:

```xml
  <configBuilders>
    <builders>
      <add name="AzureAppSettings" mode="Greedy" prefix="APPSETTING_" stripPrefix="true" type="Microsoft.Configuration.ConfigurationBuilders.EnvironmentConfigBuilder, Microsoft.Configuration.ConfigurationBuilders.Environment, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35" />
      <add name="AzureConnStr_Sql" mode="Greedy" prefix="SQLCONNSTR_" stripPrefix="true" type="Microsoft.Configuration.ConfigurationBuilders.EnvironmentConfigBuilder, Microsoft.Configuration.ConfigurationBuilders.Environment, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35" />
      <add name="AzureConnStr_MySql" mode="Greedy" prefix="MYSQLCONNSTR_" stripPrefix="true" type="Microsoft.Configuration.ConfigurationBuilders.EnvironmentConfigBuilder, Microsoft.Configuration.ConfigurationBuilders.Environment, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35" />
      <add name="AzureConnStr_SqlAzure" mode="Greedy" prefix="SQLAZURECONNSTR_" stripPrefix="true" type="Microsoft.Configuration.ConfigurationBuilders.EnvironmentConfigBuilder, Microsoft.Configuration.ConfigurationBuilders.Environment, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35" />
      <add name="AzureConnStr_Postgres" mode="Greedy" prefix="POSTGRESQLCONNSTR_" stripPrefix="true" type="Microsoft.Configuration.ConfigurationBuilders.EnvironmentConfigBuilder, Microsoft.Configuration.ConfigurationBuilders.Environment, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35" />
      <add name="AzureConnStr_Custom" mode="Greedy" prefix="CUSTOMCONNSTR_" stripPrefix="true" type="Microsoft.Configuration.ConfigurationBuilders.EnvironmentConfigBuilder, Microsoft.Configuration.ConfigurationBuilders.Environment, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35" />
    </builders>
  </configBuilders>
  <appSettings configBuilders="AzureAppSettings" />
  <connectionStrings configBuilders="AzureConnStr_Sql,AzureConnStr_MySql,AzureConnStr_SqlAzure,AzureConnStr_Postgres,AzureConnStr_Custom" />
```
Use those builders along with your own and voila! You're in business.
